cvmfs/compression.cc

```diff
@@ -779,4 +787,120 @@ bool DecompressMem2Mem(const void *buf, const int64_t size,
   return true;
 }
 
+Compressor* Compressor::Clone() {
+  return new Compressor();
+}
+
+ZlibCompressor::ZlibCompressor() {
+  stream_.zalloc   = Z_NULL;
+  stream_.zfree    = Z_NULL;
+  stream_.opaque   = Z_NULL;
+  stream_.next_in  = Z_NULL;
+  stream_.avail_in = 0;
+  const int zlib_retval = deflateInit(&stream_, Z_DEFAULT_COMPRESSION);
+  assert(zlib_retval == 0);
+}
+
+
+Compressor* ZlibCompressor::Clone() {
+  ZlibCompressor* other = new ZlibCompressor();
+  assert(stream_.avail_in == 0);
+  // Delete the other stream
+  int retcode = deflateEnd(&other->stream_);
+  assert(retcode == Z_OK);
+  retcode = deflateCopy(const_cast<z_streamp>(&other->stream_), &stream_);
+  assert(retcode == Z_OK);
+  return other;
+  
+  
+}
+
+int ZlibCompressor::Deflate(upload::CharBuffer &outbuf, size_t &outbufsize, 
+            const unsigned char* inbuf, const size_t inbufsize, 
+            const bool flush) 
+{
+  // Adding compression
+  stream_.avail_in = inbufsize;
+  stream_.next_in = const_cast<unsigned char*>(inbuf);
+  const int flush_int = (flush) ? Z_FINISH : Z_NO_FLUSH;
+  upload::CharBuffer* big_buf = new upload::CharBuffer(0);
+  int retcode = 0;
+  
+  do {
+    
+    // Create a new buffer, big enough to hold the old big_buf, and whhatever
+    // else needs to be deflated
+    size_t needed_space = DeflateBound(stream_.avail_in);
+    upload::CharBuffer* tmp_buf = new upload::CharBuffer(big_buf->used_bytes() + needed_space);
+    memcpy(tmp_buf->free_space_ptr(), big_buf->ptr(), big_buf->used_bytes());
+    tmp_buf->SetUsedBytes(big_buf->used_bytes());
+    delete big_buf;

```

Comment:
During the first iteration, you are correct.  The big_buf will have 0 used bytes.  But, in the next iteration it could have many used bytes.

-Derek

> On Oct 13, 2015, at 3:34 AM, René Meusel <notifications@github.com> wrote:
> 
> In cvmfs/compression.cc:
> 
> > +  // Adding compression
> > +  stream_.avail_in = inbufsize;
> > +  stream_.next_in = const_cast<unsigned char*>(inbuf);
> > +  const int flush_int = (flush) ? Z_FINISH : Z_NO_FLUSH;
> > +  upload::CharBuffer* big_buf = new upload::CharBuffer(0);
> > +  int retcode = 0;
> > +  
> > +  do {
> > +    
> > +    // Create a new buffer, big enough to hold the old big_buf, and whhatever
> > +    // else needs to be deflated
> > +    size_t needed_space = DeflateBound(stream_.avail_in);
> > +    upload::CharBuffer* tmp_buf = new upload::CharBuffer(big_buf->used_bytes() + needed_space);
> > +    memcpy(tmp_buf->free_space_ptr(), big_buf->ptr(), big_buf->used_bytes());
> > +    tmp_buf->SetUsedBytes(big_buf->used_bytes());
> > +    delete big_buf;
> 
> Not sure if this is correct. big_buf was just initialised as empty, how can it have any used_bytes() to be copied into tmp_buf?
> 
> —
> Reply to this email directly or view it on GitHub.
>